src/client.ts

```diff
@@ -165,95 +149,151 @@ export class SubscriptionClient {
   }
 
   public unsubscribe(id: number) {
-    delete this.subscriptions[id];
-    delete this.waitingSubscriptions[id];
-    let message = { id: id, type: SUBSCRIPTION_END};
-    this.sendMessage(message);
+    if (this.requests[id]) {
+      delete this.requests[id];
+    }
+    this.sendMessage(id, MessageTypes.GQL_END, undefined);
   }
 
   public unsubscribeAll() {
-    Object.keys(this.subscriptions).forEach( subId => {
-      this.unsubscribe(parseInt(subId));
+    Object.keys(this.requests).forEach( subId => {
+      this.unsubscribe(parseInt(subId, 10));
     });
   }
 
-  // send message, or queue it if connection is not open
-  private sendMessage(message: Object) {
-    switch (this.client.readyState) {
+  private executeRequest(options: RequestOptions, handler: (error: Error[], result?: any) => void): number {
+    const { query, variables, operationName } = options;
+
+    if (!query) {
+      throw new Error('Must provide a `query`.');
+    }
+
+    if (!handler) {
+      throw new Error('Must provide an `handler`.');
+    }
+
+    if (
+      ( !isString(query) && !getOperationAST(query, operationName)) ||
+      ( operationName && !isString(operationName)) ||
+      ( variables && !isObject(variables))
+    ) {
+      throw new Error('Incorrect option types. `query` must be a string or a document,' +
+        '`operationName` must be a string, and `variables` must be an object.');
+    }
+
+    const reqId = this.generateRequestId();
+    this.requests[reqId] = { options, handler };
+    this.sendMessage(reqId, MessageTypes.GQL_START, options);
+
+    return reqId;
+  }
+
+  private buildMessage(id: number, type: string, payload: any) {
+    const payloadToReturn = payload && payload.query ?
+      {
+        ...payload,
+        query: typeof payload.query === 'string' ? payload.query : print(payload.query),
+      } :
+      payload;
+
+    return {
+      id,
+      type,
+      payload: payloadToReturn,
+    };
+  }
 
+  // ensure we have an array of errors
+  private formatErrors(errors: any) {
+    if (Array.isArray(errors)) {
+      return errors;
+    }
+    if (errors && errors.message) {
+      return [errors];
+    }
+    return [{ message: 'Unknown error' }];
+  }
+
+  private sendMessage(id: number, type: string, payload: any) {
+    this.sendMessageRaw(this.buildMessage(id, type, payload));
+  }
+
+  // send message, or queue it if connection is not open
+  private sendMessageRaw(message: Object) {
+    switch (this.status) {
       case this.client.OPEN:
-        // TODO: throw error if message isn't json serializable?
-        this.client.send(JSON.stringify(message));
+        let serializedMessage: string = JSON.stringify(message);
+        let parsedMessage: any;
+        try {
+          parsedMessage = JSON.parse(serializedMessage);
```

Comment:
`JSON.stringify` can throw errors due to circular structures. `const a = {}; const b ={a}; a.b = b; JSON.stringify(a);`